pyfar/dsp/fractional_octave_smoothing.py

```diff
@@ -0,0 +1,317 @@
+    the array 'mean_size'.
+
+    Parameters
+    ----------
+    data : ndarray
+        Array of shape (N, M) of the data.
+        (N= Number of channels, M=Length of Signal)
+    pad_width : int
+        Padding length
+    mean_size : ndarray
+        Array with numbers of samples used to compute the mean.
+
+    Returns
+    -------
+    ndarray
+        Padded data matrix containing the data values and specified padding
+        value for each frequency bin.
+
+    Raises
+    ------
+    ValueError
+        Data array must be 2D.
+    """
+    if not data.ndim == 2:
+        raise ValueError('Data array must be 2D.')
+    padded = np.array([np.pad(data, ((0, 0), (0, pad_width)),
+                              'mean', stat_length=(m)) for m in mean_size])
+    # move channel axis to front and return
+    return np.moveaxis(padded, 1, 0)
+
+
+def polar2cartesian(amplitude, phase):
+    """
+    Converts two arrays of amplitude and phase into one array
+    of complex numbers in cartesian form.
+
+    Parameters
+    ----------
+    amplitude : ndarray
+        Array of amplitude values.
+    phase : ndarray
+        Array of phase values.
+
+    Returns
+    -------
+    ndarray
+        Array of same shape as input arrays with complex numbers.
+
+    Raises
+    ------
+    ValueError
+        Arrays must have same shape.
+    """
+    if not (amplitude.shape == phase.shape):
+        raise ValueError("Arrays must have same shapes.")
+    # Save input shape
+    input_shape = amplitude.shape
+    # Reshape for cmath rect to 1D arrays:
+    reshaped_amplitude = amplitude.reshape(1, -1)[0]
+    reshaped_phase = phase.reshape(1, -1)[0]
+    # Apply cmath rect on elements:
+    cartesian = np.array([cm.rect(a, p) for a, p in
+                          zip(reshaped_amplitude, reshaped_phase)])
+    return cartesian.reshape(input_shape)
+
+
+def frac_smooth_signal(signal, smoothing_width):
+    """
+    Method to smooth a given signal.
+    Takes the data of a given signal of shape (n, m), where n is number of
+    channels and m length of the signal.
+    Creates an object of class FractionalSmoothing to compute integration
+    limits and smoothing weights. Finally applies the weights on the input data
+    and returns smoothed signal.
+
+    Parameters
+    ----------
+    signal : Signal
+        Input signal to be smoothed
+    smoothing_width : float, int
+        Width of smoothing window relative to an octave
+
+    Returns
+    -------
+    Signal
+        Smoothed Signal
+
+    Raises
+    ------
+    TypeError
+        Input data must be of type Signal.
+    """
+    if not isinstance(signal, Signal):
+        raise TypeError("Input data must be of type Signal.")
+    # Create smoothing bject
+    obj = FractionalSmoothing(signal.n_bins, smoothing_width)
+    # Compute weights:
+    obj.calc_weights()
+    # Compute smoothed magnitude spectrum
+    data = obj.apply(signal.freq)
```

Comment:
`apply()` now takes a signal as argument. Is it necessary to use `signal.copy()` to copy the input signal into a buffer, to avoid changing the signal?
Since I am only changing the domain and fft_norm (both not mutable), I can do `signal_buffer = input_signal`, correct?

Reply:
When I tested it, the input signal seemed not to change. To make sure, can you add a test for this to your to do list. I think we already have code for that on a branch. I will look for it...